src/mz/src/bin/mz/upgrader.rs

```diff
@@ -0,0 +1,289 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Command-line version checker.
+
+use axum::http::HeaderValue;
+use hyper::{
+    header::{ACCEPT, USER_AGENT},
+    HeaderMap,
+};
+use mz::{error::Error, ui::OutputFormatter, VERSION};
+use mz_build_info::build_info;
+use serde::Deserialize;
+
+use std::io::Write;
+use std::path::PathBuf;
+use std::time::{SystemTime, UNIX_EPOCH};
+use std::{env, path::Path};
+use std::{
+    fs::{File, OpenOptions},
+    io::{BufRead, BufReader},
+};
+
+/// This const variable represents a whole week in secs =
+/// (7 days * 24 hours * 60 minutes * 60 seconds)
+/// Chronos lib could fit very well here but I avoided
+/// to add an additional lib to the crate.
+const SECONDS_IN_A_WEEK: u64 = 7 * 24 * 60 * 60;
+const TEMP_FILE_NAME: &str = ".mz.ver";
+const MZ_V_PREFIX: &str = "mz-v";
+
+pub struct UpgradeChecker {
+    no_color: bool,
+}
+
+impl UpgradeChecker {
+    pub fn new(no_color: bool) -> Self {
+        UpgradeChecker { no_color }
+    }
+
+    /// Writes the current timestmap in the temp file.
+    fn update_temp_file(&self, (major, minor, patch): (u64, u64, u64)) -> Result<(), Error> {
+        let mut temp_path: PathBuf = env::temp_dir();
+        temp_path.push(TEMP_FILE_NAME);
+
+        let mut file = File::create(&temp_path)?;
+
+        let current_time = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .map_err(|_| Error::TimestampConversionError)?
+            .as_secs();
```

Comment:
Rather than writing the time into the file, recommend just using the modification time of the file: https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.modified